hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/resource/LeakDetector.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.apache.hadoop.hdds.resource;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.ref.ReferenceQueue;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Simple general resource leak detector using {@link ReferenceQueue} and {@link java.lang.ref.WeakReference} to
+ * observe resource object life-cycle and assert proper resource closure before they are GCed.
+ *
+ * <p>
+ * Example usage:
+ *
+ * <pre> {@code
+ * class MyResource implements AutoClosable {
+ *   static final LeakDetector LEAK_DETECTOR = new LeakDetector("MyResource");
+ *
+ *   private final LeakTracker leakTracker = LEAK_DETECTOR.track(this, () -> {
+ *      // report leaks, don't refer to the original object (MyResource) here.
+ *      System.out.println("MyResource is not closed before being discarded.");
+ *   });
+ *
+ *   @Override
+ *   public void close() {
+ *     // proper resources cleanup...
+ *     // inform tracker that this object is closed properly.
+ *     leakTracker.close();
+ *   }
+ * }
+ *
+ * }</pre>
+ */
+public class LeakDetector implements Runnable {
+  public static final Logger LOG = LoggerFactory.getLogger(LeakDetector.class);
+  private final ReferenceQueue<Object> queue = new ReferenceQueue<>();
+  private final Set<LeakTracker> allLeaks = Collections.newSetFromMap(new ConcurrentHashMap<>());
```

Comment:
This set is not needed since `LeakTracker` can have a boolean to track whether it is closed.

Reply:
It doesn't work that way. The weak references need to be kept somewhere so ReferenceQueue works. The pattern of using a set to keep track of weak references is standard. Here we just also leverage the set to keep track of object closure.

Reply:
Are you sure?  Why the weak references need to be kept somewhere so ReferenceQueue works?

Reply:
Yes, it's documented [here](https://docs.oracle.com/javase/8/docs/api/java/lang/ref/package-summary.html)

>The relationship between a registered reference object and its queue is one-sided. That is, a queue does not keep track of the references that are registered with it. If a registered reference becomes unreachable itself, then it will never be enqueued. It is the responsibility of the program using reference objects to ensure that the objects remain reachable for as long as the program is interested in their referents.

We need to ensure the WeakReference object itself is reachable. (The pointer from the manage objects to the WeakReference, or `LeakTracker`, does not count).

I've referred to different leak detector implementations, including netty, to observe this pattern.

Reply:
Summary:
- Keeping WeakReference of resources to keep the reference reachable. 
- When closing resources, remove the references from the set to make them unreachable. This way, (mostly) only references of unclosed resources are enqueued. => Optimized.

Reply:
Thanks for the info!  Good to know.  Then, we will need a data structure to hold the references.